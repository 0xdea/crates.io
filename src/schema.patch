diff --git a/src/schema.rs b/src/schema.rs
index 27d068510..f2a8546aa 100644
--- a/src/schema.rs
+++ b/src/schema.rs
@@ -1,3 +1,5 @@
+#![allow(unused_imports)]
+
 table! {
     use diesel::sql_types::*;
     use diesel_full_text_search::{TsVector as Tsvector};
@@ -183,12 +185,6 @@ table! {
         ///
         /// (Automatically generated by Diesel.)
         created_at -> Timestamp,
-        /// The `path` column of the `categories` table.
-        ///
-        /// Its SQL type is `Ltree`.
-        ///
-        /// (Automatically generated by Diesel.)
-        path -> Ltree,
     }
 }
 
@@ -701,6 +697,24 @@ table! {
     }
 }
 
+table! {
+    /// Representation of the `recent_crate_downloads` view.
+    ///
+    /// This data represents the downloads in the last 90 days.
+    /// This view does not contain realtime data.
+    /// It is refreshed by the `update-downloads` script.
+    recent_crate_downloads (crate_id) {
+        /// The `crate_id` column of the `recent_crate_downloads` view.
+        ///
+        /// Its SQL type is `Integer`.
+        crate_id -> Integer,
+        /// The `downloads` column of the `recent_crate_downloads` table.
+        ///
+        /// Its SQL type is `BigInt`.
+        downloads -> BigInt,
+    }
+}
+
 table! {
     use diesel::sql_types::*;
     use diesel_full_text_search::{TsVector as Tsvector};
@@ -1028,7 +1042,8 @@ joinable!(api_tokens -> users (user_id));
 joinable!(badges -> crates (crate_id));
 joinable!(crate_owner_invitations -> crates (crate_id));
 joinable!(crate_owners -> crates (crate_id));
-joinable!(crate_owners -> users (created_by));
+joinable!(crate_owners -> teams (owner_id));
+joinable!(crate_owners -> users (owner_id));
 joinable!(crates_categories -> categories (category_id));
 joinable!(crates_categories -> crates (crate_id));
 joinable!(crates_keywords -> crates (crate_id));
@@ -1041,6 +1056,7 @@ joinable!(follows -> users (user_id));
 joinable!(publish_limit_buckets -> users (user_id));
 joinable!(publish_rate_overrides -> users (user_id));
 joinable!(readme_renderings -> versions (version_id));
+joinable!(recent_crate_downloads -> crates (crate_id));
 joinable!(version_downloads -> versions (version_id));
 joinable!(version_owner_actions -> api_tokens (api_token_id));
 joinable!(version_owner_actions -> users (user_id));
@@ -1067,6 +1083,7 @@ allow_tables_to_appear_in_same_query!(
     publish_limit_buckets,
     publish_rate_overrides,
     readme_renderings,
+    recent_crate_downloads,
     reserved_crate_names,
     teams,
     users,
