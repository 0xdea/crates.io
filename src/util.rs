pub use self::bytes_request::BytesRequest;
pub use self::io_util::{read_fill, read_le_u32};
pub use self::request_helpers::*;

mod bytes_request;
pub mod diesel;
pub mod errors;
mod io_util;
mod request_helpers;
pub mod rfc3339;
pub mod token;
pub mod tracing;
