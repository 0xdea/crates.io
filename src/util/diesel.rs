use diesel::connection::LoadConnection;
use diesel::pg::Pg;
use diesel::result::Error;

pub trait Conn: LoadConnection<Backend = Pg> {}

impl<T> Conn for T where T: LoadConnection<Backend = Pg> {}

pub fn is_read_only_error(error: &Error) -> bool {
    matches!(error, Error::DatabaseError(_, info) if info.message().ends_with("read-only transaction"))
}
