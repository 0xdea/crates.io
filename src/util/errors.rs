use std::io::{IoError, MemReader};
use std::fmt;
use std::fmt::{Show, Formatter, FormatError};
use std::collections::HashMap;

use conduit::Response;
use curl::ErrCode;
use pg::error::PostgresError;
use serialize::json;

pub trait CargoError: Send {
    fn description(&self) -> String;
    fn detail(&self) -> Option<String> { None }
    fn cause<'a>(&'a self) -> Option<&'a CargoError + Send> { None }

    fn to_error<E: FromError<Self>>(self) -> E {
        FromError::from_error(self)
    }

    fn box_error(self) -> Box<CargoError + Send> {
        box self as Box<CargoError + Send>
    }

    fn concrete(&self) -> ConcreteCargoError {
        ConcreteCargoError {
            description: self.description(),
            detail: self.detail(),
            cause: self.cause().map(|c| box c.concrete() as Box<CargoError + Send>),
        }
    }

    fn with_cause<E: CargoError + Send>(self, cause: E) -> Box<CargoError + Send> {
        let mut concrete = self.concrete();
        concrete.cause = Some(cause.box_error());
        box concrete as Box<CargoError + Send>
    }

    fn response(&self) -> Option<Response> { None }
}

pub trait FromError<E> {
    fn from_error(error: E) -> Self;
}

impl<E: CargoError + Send> FromError<E> for Box<CargoError + Send> {
    fn from_error(error: E) -> Box<CargoError + Send> {
        error.box_error()
    }
}

macro_rules! from_error (
    ($ty:ty) => {
        impl FromError<$ty> for $ty {
            fn from_error(error: $ty) -> $ty {
                error
            }
        }
    }
)

impl Show for Box<CargoError + Send> {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        write!(f, "{}", self.description())
    }
}

impl CargoError for Box<CargoError + Send> {
    fn description(&self) -> String {
        (*self).description()
    }

    fn detail(&self) -> Option<String> {
        (*self).detail()
    }

    fn cause<'a>(&'a self) -> Option<&'a CargoError + Send> {
        (*self).cause()
    }

    fn box_error(self) -> Box<CargoError + Send> {
        self
    }
}

pub type CargoResult<T> = Result<T, Box<CargoError + Send>>;

pub trait BoxError<T> {
    fn box_error(self) -> CargoResult<T>;
}

pub trait ChainError<T> {
    fn chain_error<E: CargoError + Send>(self, callback: || -> E) -> CargoResult<T> ;
}

impl<'a, T> ChainError<T> for ||:'a -> CargoResult<T> {
    fn chain_error<E: CargoError + Send>(self, callback: || -> E) -> CargoResult<T> {
        self().map_err(|err| callback().with_cause(err))
    }
}

impl<T, E: CargoError + Send> BoxError<T> for Result<T, E> {
    fn box_error(self) -> CargoResult<T> {
        self.map_err(|err| err.box_error())
    }
}

impl<T, E: CargoError + Send> ChainError<T> for Result<T, E> {
    fn chain_error<E: CargoError + Send>(self, callback: || -> E) -> CargoResult<T>  {
        self.map_err(|err| callback().with_cause(err))
    }
}

impl CargoError for IoError {
    fn description(&self) -> String { self.to_string() }
}

from_error!(IoError)

impl CargoError for FormatError {
    fn description(&self) -> String {
        "formatting failed".to_string()
    }
}

from_error!(FormatError)

impl CargoError for PostgresError {
    fn description(&self) -> String { self.to_string() }
}

from_error!(PostgresError)

impl CargoError for ErrCode {
    fn description(&self) -> String { self.to_string() }
}

from_error!(ErrCode)

impl CargoError for json::DecoderError {
    fn description(&self) -> String { self.to_string() }
}

from_error!(json::DecoderError)

pub struct ConcreteCargoError {
    description: String,
    detail: Option<String>,
    cause: Option<Box<CargoError + Send>>,
}

impl Show for ConcreteCargoError {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        write!(f, "{}", self.description)
    }
}

impl CargoError for ConcreteCargoError {
    fn description(&self) -> String {
        self.description.clone()
    }

    fn detail(&self) -> Option<String> {
        self.detail.clone()
    }

    fn cause<'a>(&'a self) -> Option<&'a CargoError + Send> {
        self.cause.as_ref().map(|c| { let err: &CargoError + Send = *c; err })
    }

    fn with_cause<E: CargoError + Send>(mut self,
                                        err: E) -> Box<CargoError + Send> {
        self.cause = Some(err.box_error());
        box self as Box<CargoError + Send>
    }
}

pub struct NotFound;

impl CargoError for NotFound {
    fn description(&self) -> String { "not found".to_string() }

    fn response(&self) -> Option<Response> {
        Some(Response {
            status: (404, "Not Found"),
            headers: HashMap::new(),
            body: box MemReader::new(Vec::new()),
        })
    }
}

from_error!(NotFound)

pub struct Unauthorized;

impl CargoError for Unauthorized {
    fn description(&self) -> String { "unauthorized".to_string() }

    fn response(&self) -> Option<Response> {
        Some(Response {
            status: (403, "Forbidden"),
            headers: HashMap::new(),
            body: box MemReader::new(Vec::new()),
        })
    }
}

from_error!(Unauthorized)

#[allow(dead_code)]
pub fn internal_error<S1: Str, S2: Str>(error: S1,
                                        detail: S2) -> Box<CargoError + Send> {
    box ConcreteCargoError {
        description: error.as_slice().to_string(),
        detail: Some(detail.as_slice().to_string()),
        cause: None,
    } as Box<CargoError + Send>
}

pub fn internal<S: Show>(error: S) -> Box<CargoError + Send> {
    box ConcreteCargoError {
        description: error.to_string(),
        detail: None,
        cause: None,
    } as Box<CargoError + Send>
}
