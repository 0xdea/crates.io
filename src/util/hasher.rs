use std::io;
use openssl::crypto::hash::{Hasher, HashType};

pub struct HashingReader<R> {
    inner: R,
    hasher: Hasher,
}

impl<R: Reader> HashingReader<R> {
    pub fn new(r: R) -> HashingReader<R> {
        HashingReader { inner: r, hasher: Hasher::new(HashType::SHA256) }
    }

    pub fn finalize(self) -> Vec<u8> { self.hasher.finalize() }
}

impl<R: Reader> Reader for HashingReader<R> {
    fn read(&mut self, buf: &mut [u8]) -> io::IoResult<uint> {
        let amt = try!(self.inner.read(buf));
        self.hasher.update(buf.slice_to(amt));
        return Ok(amt)
    }
}
