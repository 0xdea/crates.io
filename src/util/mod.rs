use std::io::{MemReader, IoError};
use std::io::process::{ProcessOutput, Command};
use std::collections::HashMap;
use std::fmt::Show;
use std::str;

use serialize::{json, Encodable};
use url;

use conduit::{Request, Response, Handler};

pub use self::errors::{CargoError, CargoResult, internal, human, internal_error};
pub use self::errors::{ChainError, BoxError};
pub use self::result::{Require, Wrap};
pub use self::lazy_cell::LazyCell;
pub use self::io::LimitErrorReader;
pub use self::hasher::HashingReader;

pub mod errors;
pub mod result;
mod lazy_cell;
mod io;
mod hasher;

pub trait RequestUtils {
    fn redirect(self, url: String) -> Response;

    fn json<'a, T: Encodable<json::Encoder<'a>, IoError>>(self, t: &T) -> Response;
    fn query(self) -> HashMap<String, String>;
    fn wants_json(self) -> bool;
}

pub fn json_response<'a, T>(t: &T) -> Response
                            where T: Encodable<json::Encoder<'a>, IoError> {
    let s = json::encode(t);
    let mut headers = HashMap::new();
    headers.insert("Content-Type".to_string(),
                   vec!["application/json; charset=utf-8".to_string()]);
    Response {
        status: (200, "OK"),
        headers: headers,
        body: box MemReader::new(s.into_bytes()),
    }
}


impl<'a> RequestUtils for &'a Request + 'a {
    fn json<'a, T: Encodable<json::Encoder<'a>, IoError>>(self, t: &T) -> Response {
        json_response(t)
    }

    fn query(self) -> HashMap<String, String> {
        url::form_urlencoded::parse_str(self.query_string().unwrap_or(""))
            .into_iter().collect()
    }

    fn redirect(self, url: String) -> Response {
        let mut headers = HashMap::new();
        headers.insert("Location".to_string(), vec![url.to_string()]);
        Response {
            status: (302, "Found"),
            headers: headers,
            body: box MemReader::new(Vec::new()),
        }
    }

    fn wants_json(self) -> bool {
        let content = self.headers().find("Accept").unwrap_or(Vec::new());
        content.iter().any(|s| s.contains("json"))
    }
}

pub struct C(pub fn(&mut Request) -> CargoResult<Response>);

impl Handler for C {
    fn call(&self, req: &mut Request) -> Result<Response, Box<Show + 'static>> {
        let C(f) = *self;
        match f(req) {
            Ok(req) => Ok(req),
            Err(e) => {
                match e.response() {
                    Some(response) => Ok(response),
                    None => Err(box e as Box<Show>),
                }
            }
        }
    }
}

pub fn exec(cmd: &Command) -> CargoResult<ProcessOutput> {
    let output = try!(cmd.output().chain_error(|| {
        internal(format!("failed to run command `{}`", cmd))
    }));
    if !output.status.success() {
        let mut desc = String::new();
        if output.output.len() != 0 {
            desc.push_str("--- stdout\n");
            desc.push_str(str::from_utf8(output.output.as_slice()).unwrap());
        }
        if output.error.len() != 0 {
            desc.push_str("--- stderr\n");
            desc.push_str(str::from_utf8(output.error.as_slice()).unwrap());
        }
        Err(internal_error(format!("failed to run command `{}`", cmd), desc))
    } else {
        Ok(output)
    }
}
