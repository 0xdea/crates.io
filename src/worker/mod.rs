//! The `worker` module contains all the tasks that can be queued up for the
//! background worker process to work on. This includes recurring tasks like
//! the daily database maintenance, but also operations like rendering READMEs
//! and uploading them to S3.

pub mod cloudfront;
mod daily_db_maintenance;
pub mod dump_db;
pub mod fastly;
mod git;
mod readmes;
mod update_downloads;

pub(crate) use daily_db_maintenance::perform_daily_db_maintenance;
pub(crate) use dump_db::DumpDbJob;
pub(crate) use git::{
    perform_index_squash, sync_to_git_index, sync_to_sparse_index, NormalizeIndexJob,
    SyncToIndexJob,
};
pub(crate) use readmes::RenderAndUploadReadmeJob;
pub(crate) use update_downloads::perform_update_downloads;
