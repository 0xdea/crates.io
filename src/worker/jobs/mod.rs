use crate::util::diesel::Conn;
use crates_io_worker::{BackgroundJob, EnqueueError};
use std::fmt::Display;

mod archive_version_downloads;
mod daily_db_maintenance;
mod delete_crate;
mod downloads;
pub mod dump_db;
mod expiry_notification;
mod index;
mod index_version_downloads_archive;
mod readmes;
pub mod rss;
mod send_publish_notifications;
mod sync_admins;
mod typosquat;
mod update_default_version;

pub use self::archive_version_downloads::ArchiveVersionDownloads;
pub use self::daily_db_maintenance::DailyDbMaintenance;
pub use self::delete_crate::DeleteCrateFromStorage;
pub use self::downloads::{
    CleanProcessedLogFiles, ProcessCdnLog, ProcessCdnLogQueue, UpdateDownloads,
};
pub use self::dump_db::DumpDb;
pub use self::expiry_notification::SendTokenExpiryNotifications;
pub use self::index::{NormalizeIndex, SquashIndex, SyncToGitIndex, SyncToSparseIndex};
pub use self::index_version_downloads_archive::IndexVersionDownloadsArchive;
pub use self::readmes::RenderAndUploadReadme;
pub use self::send_publish_notifications::SendPublishNotificationsJob;
pub use self::sync_admins::SyncAdmins;
pub use self::typosquat::CheckTyposquat;
pub use self::update_default_version::UpdateDefaultVersion;

/// Enqueue both index sync jobs (git and sparse) for a crate, unless they
/// already exist in the background job queue.
///
/// Note that there are currently no explicit tests for this functionality,
/// since our test suite only allows us to use a single database connection
/// and the background worker queue locking only work when using multiple
/// connections.
#[instrument(name = "swirl.enqueue", skip_all, fields(message = "sync_to_index", krate = %krate))]
pub fn enqueue_sync_to_index<T: Display>(
    krate: T,
    conn: &mut impl Conn,
) -> Result<(), EnqueueError> {
    let krate = krate.to_string();

    SyncToGitIndex::new(krate.clone()).enqueue(conn)?;
    SyncToSparseIndex::new(krate).enqueue(conn)?;

    Ok(())
}
