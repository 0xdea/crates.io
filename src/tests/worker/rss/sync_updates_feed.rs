use crate::schema::{crates, versions};
use crate::tests::util::TestApp;
use crate::worker::jobs;
use chrono::DateTime;
use crates_io_worker::BackgroundJob;
use diesel::prelude::*;
use diesel::{PgConnection, RunQueryDsl};
use insta::assert_snapshot;

#[tokio::test(flavor = "multi_thread")]
async fn test_sync_updates_feed() {
    let (app, _) = TestApp::full().empty();

    app.db(|conn| {
        let d = "let's try & break this <item>";

        create_version(conn, "foo", "0.1.0", None, "2024-06-20T10:13:54Z");
        create_version(conn, "foo", "0.1.1", None, "2024-06-20T12:45:12Z");
        create_version(conn, "foo", "1.0.0", None, "2024-06-21T17:01:33Z");
        create_version(conn, "bar", "3.0.0-beta.1", Some(d), "2024-06-21T17:03:45Z");
        create_version(conn, "foo", "1.1.0", None, "2024-06-22T08:30:01Z");
        create_version(conn, "foo", "1.2.0", None, "2024-06-22T15:57:19Z");

        jobs::rss::SyncUpdatesFeed.enqueue(conn).unwrap();
    });

    app.run_pending_background_jobs().await;

    assert_snapshot!(app.stored_files().await.join("\n"), @"rss/updates.xml");

    let store = app.as_inner().storage.as_inner();
    let result = store.get(&"rss/updates.xml".into()).await.unwrap();
    let bytes = result.bytes().await.unwrap();
    let content = String::from_utf8(bytes.to_vec()).unwrap();
    assert_snapshot!(content);
}

fn create_version(
    conn: &mut PgConnection,
    name: &str,
    version: &str,
    description: Option<&str>,
    publish_time: &str,
) -> i32 {
    let publish_time = DateTime::parse_from_rfc3339(publish_time)
        .unwrap()
        .naive_utc();

    let crate_id = crates::table
        .select(crates::id)
        .filter(crates::name.eq(name))
        .get_result::<i32>(conn)
        .optional()
        .unwrap();

    let crate_id = crate_id.unwrap_or_else(|| {
        diesel::insert_into(crates::table)
            .values((
                crates::name.eq(name),
                crates::description.eq(description),
                crates::created_at.eq(publish_time),
                crates::updated_at.eq(publish_time),
            ))
            .returning(crates::id)
            .get_result(conn)
            .unwrap()
    });

    diesel::insert_into(versions::table)
        .values((
            versions::crate_id.eq(crate_id),
            versions::num.eq(version),
            versions::created_at.eq(publish_time),
            versions::updated_at.eq(publish_time),
            versions::checksum.eq("checksum"),
        ))
        .returning(versions::id)
        .get_result(conn)
        .unwrap()
}
