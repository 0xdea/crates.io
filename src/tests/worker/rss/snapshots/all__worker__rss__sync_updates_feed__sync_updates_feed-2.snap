---
source: src/tests/worker/rss/sync_updates_feed.rs
expression: content
---
<?xml version="1.0" encoding="utf-8"?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom">
    <channel>
        <title>crates.io: recent updates</title>
        <link>https://crates.io/</link>
        <description>Recent version publishes on the crates.io package registry</description>
        <language>en</language>
        <atom:link href="https://static.crates.io/rss/updates.xml" rel="self" type="application/rss+xml"/>
        <item>
            <title>foo v1.2.0</title>
            <link>https://crates.io/crates/foo/1.2.0</link>
            <guid>https://crates.io/crates/foo/1.2.0</guid>
            <pubDate>Sat, 22 Jun 2024 15:57:19 +0000</pubDate>
        </item>
        <item>
            <title>foo v1.1.0</title>
            <link>https://crates.io/crates/foo/1.1.0</link>
            <guid>https://crates.io/crates/foo/1.1.0</guid>
            <pubDate>Sat, 22 Jun 2024 08:30:01 +0000</pubDate>
        </item>
        <item>
            <title>bar v3.0.0-beta.1</title>
            <link>https://crates.io/crates/bar/3.0.0-beta.1</link>
            <description><![CDATA[let&apos;s try &amp; break this &lt;item&gt;]]></description>
            <guid>https://crates.io/crates/bar/3.0.0-beta.1</guid>
            <pubDate>Fri, 21 Jun 2024 17:03:45 +0000</pubDate>
        </item>
        <item>
            <title>foo v1.0.0</title>
            <link>https://crates.io/crates/foo/1.0.0</link>
            <guid>https://crates.io/crates/foo/1.0.0</guid>
            <pubDate>Fri, 21 Jun 2024 17:01:33 +0000</pubDate>
        </item>
        <item>
            <title>foo v0.1.1</title>
            <link>https://crates.io/crates/foo/0.1.1</link>
            <guid>https://crates.io/crates/foo/0.1.1</guid>
            <pubDate>Thu, 20 Jun 2024 12:45:12 +0000</pubDate>
        </item>
        <item>
            <title>foo v0.1.0</title>
            <link>https://crates.io/crates/foo/0.1.0</link>
            <guid>https://crates.io/crates/foo/0.1.0</guid>
            <pubDate>Thu, 20 Jun 2024 10:13:54 +0000</pubDate>
        </item>
    </channel>
</rss>
